### PR TITLE
Ignore compiler and STL version mismatch on published packages to unblock ApiView

### PR DESCRIPTION
**Pull request by LarryOsterman**
The current MSVC STL version requires clang 16, but the newest version of clang in VCPKG is clang 15. This change disables the compiler STL version checks to allow the ApiView tool to continue to work with the newest Microsoft STL version.

# Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/main/CONTRIBUTING.md#pull-requests).

- [ ] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [ ] Doxygen docs
- [ ] Unit tests
- [ ] No unwanted commits/changes
- [ ] Descriptive title/description
  - [ ] PR is single purpose
  - [ ] Related issue listed
- [ ] Comments in source
- [ ] No typos
- [ ] Update changelog
- [ ] Not work-in-progress
- [ ] External references or docs updated
- [ ] Self review of PR done
- [ ] Any breaking changes?

### CODE REVIEW

File: sdk/attestation/azure-security-attestation/inc/ApiViewSettings.json [hunk 1]
```diff
@@ -5,7 +5,9 @@
   ],
   "sourceFilesToSkip": [
   ],
-  "additionalCompilerSwitches": [],
+  "additionalCompilerSwitches": [
+    "-D_ALLOW_COMPILER_AND_STL_VERSION_MISMATCH"
+  ],
```
Comment (ahsonkhan) [inline sdk/attestation/azure-security-attestation/inc/ApiViewSettings.json:10]: @CaseyCarter please review.

I vaguely recall @BillyONeal, previously, suggesting we avoid this sort of mismatch. @LarryOsterman if this is a temporary solution, what's the long term fix?

Comment (LarryOsterman) [reply sdk/attestation/azure-security-attestation/inc/ApiViewSettings.json:10]: To the best of my knowledge, there is no long-term fix. The Microsoft Visual Studio keeps requiring newer and newer clang versions, which breaks the ApiView tool. I am almost certain that those reasons have nothing to do with parts of the STL in C++14, but are related to newer C++ features.

Also, I strongly suspect that this definition is required for runtime functionality, but this is for the APiView tool which just has to be able to parse the code.

We can move this definition into the tool, but the current ApiView tool has a breaking change that has not yet been integrated to our pipeline.

Comment (LarryOsterman) [reply sdk/attestation/azure-security-attestation/inc/ApiViewSettings.json:10]: Also note: This issue will block our ability to release any packages in September, so it's critical to resolve this quickly.

Comment (BillyONeal) [reply sdk/attestation/azure-security-attestation/inc/ApiViewSettings.json:10]: >I am almost certain that those reasons have nothing to do with parts of the STL in C++14, but are related to newer C++ features.

Not always. That gets bumped when the library has bug workarounds removed too. It would be good if a current microsoft/STL maintainer would comment.

Comment (CaseyCarter) [reply sdk/attestation/azure-security-attestation/inc/ApiViewSettings.json:10]: > Also note: This issue will block our ability to release any packages in September, so it's critical to resolve this quickly.

Defining `_ALLOW_COMPILER_AND_STL_VERSION_MISMATCH` will probably work for now, and mostly continue working in the future, but we should probably have some kind of plan in place for the day that it doesn't work. If our ability to release is predicated on the use of a tool we're sometimes unable to update then I guess we just cross our fingers and hope it never breaks?

> That gets bumped when the library has bug workarounds removed too.

The vast majority (99.9%?) of Clang bug workarounds in "old" C++14 code are going to be for runtime behavior. They won't affect libclang's ability to parse the code. It's possible that libclang will be confused by some new syntactic structure in `#ifdef`ed-out "new" code, but again extremely unlikely.

**Review by ahsonkhan — APPROVED** [review]
Approving to avoid blocking the release, especially since this is about the ApiView tool and not how we build the SDK itself.

**Review by RickWinter — APPROVED** [review]

File: sdk/attestation/azure-security-attestation/inc/ApiViewSettings.json (hunk 1 above)
Comment (LarryOsterman) [reply sdk/attestation/azure-security-attestation/inc/ApiViewSettings.json:10]: FWIW, the version bump came in: https://github.com/microsoft/STL/commit/a62109595b6d89e08172fdf4beb75a2670fe0cc9

Comment (CaseyCarter) [reply sdk/attestation/azure-security-attestation/inc/ApiViewSettings.json:10]: FWIW, I don't expect there to be any significant changes to the C++14 portions of the STL that would cause breakage. The last such significant change was maybe two years ago when we removed the non-`if constexpr` codepaths.

The major compilers implement the C++17 feature `if constexpr` in earlier language modes with a suppressible warning. The STL maintained both `if constexpr` and non-`if constexpr` versions of some facilities for a few years until all of our compilers supported such an extension. There were some pains during the transition for users of libclang-based tools based on very old versions of LLVM.

TLDR: If your tool needs to parse the STL, it needs to understand C++ as well as a current C++ compiler.
